### PR TITLE
KAFKA-15853: Refactor ShareGroupConfig with AbstractConfig

### PR DESCRIPTION
**Pull request by OmniaGM**
Move getters and validation out of KafkaConfig and into ShareGroupConfig
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by OmniaGM** [comment]
@chia7712 can you have a look please whenever you have time? I started with this as the KIP-932 is under development at the moment and I don't want to clash with it

**Comment by chia7712** [comment]
@OmniaGM that refactor is good, but maybe we can reach more consensus before refactor. Please take a look at https://github.com/apache/kafka/pull/16458#discussion_r1662734441 

It would be nice we can build a pattern for all xxxConfig files.

**Comment by chia7712** [comment]
@OmniaGM as the summary of https://github.com/apache/kafka/pull/16458#issuecomment-2206822223, if the `ShareGroupConfig` is not dynamic we can introduce local attributes and validate all of them in construction. That can make sure each `ShareGroupConfig` is totally valid.

**Comment by OmniaGM** [comment]
> @OmniaGM as the summary of [#16458 (comment)](https://github.com/apache/kafka/pull/16458#issuecomment-2206822223), if the `ShareGroupConfig` is not dynamic we can introduce local attributes and validate all of them in construction. That can make sure each `ShareGroupConfig` is totally valid.

Just pushed a refactor here https://github.com/apache/kafka/pull/16506/commits/5b4b86286c3e5f240336ebf45ad6596f787e5b1d to introduce attributes and move the validation to the construction. Please have a look when you have time.

File: server/src/main/java/org/apache/kafka/server/config/ShareGroupConfig.java [hunk 1]
```diff
@@ -100,4 +102,124 @@ public class ShareGroupConfigs {
...
+        shareGroupMaxHeartbeatIntervalMs = config.getInt(ShareGroupConfig.SHARE_GROUP_MAX_HEARTBEAT_INTERVAL_MS_CONFIG);
+        shareGroupRecordLockDurationMs = config.getInt(ShareGroupConfig.SHARE_GROUP_RECORD_LOCK_DURATION_MS_CONFIG);
+        shareGroupMaxRecordLockDurationMs = config.getInt(ShareGroupConfig.SHARE_GROUP_MAX_RECORD_LOCK_DURATION_MS_CONFIG);
+        shareGroupMinRecordLockDurationMs = config.getInt(ShareGroupConfig.SHARE_GROUP_MIN_RECORD_LOCK_DURATION_MS_CONFIG);
+        shareGroupMaxSize = config.getShort(ShareGroupConfig.SHARE_GROUP_MAX_SIZE_CONFIG);
+        validate();
+    }
+
+    /** Share group configuration **/
+    Boolean isShareGroupEnabled() {
```
Comment (chia7712) [inline server/src/main/java/org/apache/kafka/server/config/ShareGroupConfig.java]: `boolean`

File: server/src/main/java/org/apache/kafka/server/config/ShareGroupConfig.java [hunk 2]
```diff
@@ -100,4 +102,124 @@ public class ShareGroupConfigs {
             .define(SHARE_GROUP_MAX_HEARTBEAT_INTERVAL_MS_CONFIG, INT, SHARE_GROUP_MAX_HEARTBEAT_INTERVAL_MS_DEFAULT, atLeast(1), MEDIUM, SHARE_GROUP_MAX_HEARTBEAT_INTERVAL_MS_DOC)
             .define(SHARE_GROUP_MAX_GROUPS_CONFIG, SHORT, SHARE_GROUP_MAX_GROUPS_DEFAULT, between(1, 100), MEDIUM, SHARE_GROUP_MAX_GROUPS_DOC)
             .define(SHARE_GROUP_MAX_SIZE_CONFIG, SHORT, SHARE_GROUP_MAX_SIZE_DEFAULT, between(10, 1000), MEDIUM, SHARE_GROUP_MAX_SIZE_DOC);
+
+    private final Boolean isShareGroupEnabled;
```
Comment (chia7712) [inline server/src/main/java/org/apache/kafka/server/config/ShareGroupConfig.java]: `boolean`

**Review by chia7712 — COMMENTED** [review]
@OmniaGM thanks for this patch. only two small comments :smile:

**Review by AndrewJSchofield — APPROVED** [review]
lgtm. I was wondering how the refactoring that @OmniaGM is doing for config would affect the work we are doing on KIP-932.

**Comment by OmniaGM** [comment]
> lgtm. I was wondering how the refactoring that @OmniaGM is doing for config would affect the work we are doing on KIP-932.

@AndrewJSchofield I believe it shouldn't have significant effect on KIP-932's work apart from the fact that now all getters, validation and config definitions will be defined in `ShareGroupConfig`. Specially that as far as I can see we still didn't define all broker configs defined in KIP-932 yet. This also should reduce conflicts between refactoring `KafkaConfig` and broker config part of KIP-932. 

Small suggestion you can depends on `ShareGroupConfig` instead of the full `KafkaConfig` where config is needed now if you wish.

**Review by chia7712 — APPROVED** [review]

**Comment by apoorvmittal10** [comment]
Thanks for the PR @OmniaGM, I do see `GroupCoordinatorConfig` also follows new way for configs 👍

**Pull request merged**